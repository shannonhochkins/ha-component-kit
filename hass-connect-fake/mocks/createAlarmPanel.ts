import { HassEntity } from 'home-assistant-js-websocket';
import { createEntity } from './createEntity';

const now = new Date();

export const defaults = {
  entity_id: "alarm_control_panel.home_alarm",
  state: "disarmed",
  attributes: {
    code_format: "number",
    changed_by: null,
    code_arm_required: true,
    supported_features: 31,
    friendly_name: "Home Alarm",
    icon: "mdi:security",
  },
  context: {
    id: "01H989071CWR02B5HG861JXWCC",
    parent_id: null,
    user_id: null,
  },
  last_changed: now.toISOString(),
  last_updated: now.toISOString(),
} satisfies HassEntity;

export const createAlarmPanel = (entity_id: string, overrides: Partial<HassEntity> = {}) => {
  return createEntity(entity_id, defaults, overrides);
}