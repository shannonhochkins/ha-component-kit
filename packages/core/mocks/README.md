This file holds the mocked entities to use with storybook, obviously the storybook isn't a direct connection to a home assistant instance, so we have to mock the entities and connections for the api.

If you want to add a new entity and the domain isn't supported, it's pretty simple! Just follow the same format as the other entities and add it to the mockEntities file, if you're unsure of the available properties for the HassEntity, you can console log the entity once you've connected to the api, however keep in mind that the properties on the entity at the time, may not be everything that's actually available!

Additionally, the api that's also mocked (HassConnectFake.tsx), does NOT contain updates for every available action, this is time consuming and sometimes irrelevant for storybook to render and "preview" the component, however if you want to indicate what the component is capable of, you can follow the same path as some of the other domains that have been mocked, eg "climate" or "light".

turnOn, turnOff and toggle have been mocked by default but this is it.