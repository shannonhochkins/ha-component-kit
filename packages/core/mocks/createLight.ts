import { HassEntity } from 'home-assistant-js-websocket';
import { createEntity } from './createEntity';

const now = new Date();

const defaults = {
  attributes: {
    friendly_name: "Dining room light",
    icon: "mdi:light-recessed",
    min_color_temp_kelvin: 2702,
    max_color_temp_kelvin: 6535,
    min_mireds: 153,
    max_mireds: 370,
    effect_list: [
      "Night",
      "Read",
      "Meeting",
      "Leasure",
      "Soft",
      "Rainbow",
      "Shine",
      "Beautiful",
      "Music",
    ],
    supported_color_modes: ["color_temp", "hs"],
    color_mode: "hs",
    brightness: 255,
    hs_color: [9, 67.1],
    rgb_color: [255, 109, 83],
    xy_color: [0.591, 0.328],
    raw_state: true,
    supported_features: 23,
  },
  state: "on",
  entity_id: "light.fake_light_1",
  last_changed: now.toISOString(),
  last_updated: now.toISOString(),
  context: {
    id: "",
    user_id: null,
    parent_id: null,
  },
} satisfies HassEntity;

export const createLight = (entity_id: string, overrides: Partial<HassEntity> = {}) => {
  return createEntity(entity_id, defaults, overrides);
}