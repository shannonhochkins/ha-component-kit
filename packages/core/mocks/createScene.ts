import { HassEntity } from 'home-assistant-js-websocket';
import { createEntity } from './createEntity';

const now = new Date();

const defaults = {
  entity_id: "scene.goodmorning",
  state: now.toISOString(),
  attributes: {
    entity_id: ["light.all_office_lights_2"],
    id: "1688447567040",
    icon: "mdi:sun-clock",
    friendly_name: "Goodmorning",
  },
  context: {
    id: "01H4FP0HY3K7R49GJS574K5TPW",
    parent_id: null,
    user_id: null,
  },
  last_changed: now.toISOString(),
  last_updated: now.toISOString(),
} satisfies HassEntity;

export const createScene = (entity_id: string, overrides: Partial<HassEntity> = {}) => {
  return createEntity(entity_id, defaults, overrides);
}