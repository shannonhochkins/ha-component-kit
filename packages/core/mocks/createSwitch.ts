import { HassEntity } from 'home-assistant-js-websocket';
import { createEntity } from './createEntity';

const now = new Date();

const defaults = {
  attributes: {
    friendly_name: "Gaming Computer",
  },
  state: "off",
  entity_id: "switch.fake_switch",
  last_changed: now.toISOString(),
  last_updated: now.toISOString(),
  context: {
    id: "",
    user_id: null,
    parent_id: null,
  },
} satisfies HassEntity;

export const createSwitch = (entity_id: string, overrides: Partial<HassEntity> = {}) => {
  return createEntity(entity_id, defaults, overrides);
}