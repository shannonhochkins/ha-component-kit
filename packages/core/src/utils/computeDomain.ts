import { AllDomains, EntityName, CamelToSnake } from "@typings";
import { snakeCase } from "lodash";

export const computeDomain = <E extends EntityName | "unknown">(entityId: E): CamelToSnake<AllDomains> =>
  snakeCase(entityId.substr(0, entityId.indexOf("."))) as CamelToSnake<AllDomains>;
