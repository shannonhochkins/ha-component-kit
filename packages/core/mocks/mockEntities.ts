import type { HassEntities } from "home-assistant-js-websocket";
import { createLight } from "./createLight";
import { createSwitch } from "./createSwitch";
import { createMediaPlayer } from './createMediaPlayer';
import { createSensor } from './createSensor';
import { createScene } from './createScene';
import { createWeather } from './createWeather';
import { createClimate } from './createClimate';
import { createVacuum } from './createVacuum';

export const entities: HassEntities = {
  ...createLight("light.fake_light_1", {
    attributes: {
      friendly_name: "Dining room light",
    },
  }),
  ...createLight("light.fake_light_2", {
    attributes: {
      friendly_name: "Office Down light",
      hs_color: undefined,
      rgb_color: undefined,
      color_mode: "color_temp",
      color_temp_kelvin: 3000,
    },
  }),
  ...createLight("light.fake_light_3", {
    attributes: {
      friendly_name: "Office striplight light",
      icon: "mdi:led-strip-variant",
      hs_color: [131, 100],
      rgb_color: [64, 255, 112],
    },
  }),
  ...createSwitch('switch.fake_switch'),
  ...createSwitch('switch.coffee_machine', {
    attributes: {
      friendly_name: "Coffee Machine",
    }
  }),
  ...createMediaPlayer('media_player.fake_tv'),
  ...createScene('scene.goodmorning'),
  ...createSensor('sensor.time', {
    attributes: {
      icon: "mdi:clock",
      friendly_name: "Time",
    },
  }),
  ...createSensor('sensor.date', {
    state: "2023-07-19",
    attributes: {
      icon: "mdi:calendar",
      friendly_name: "Date",
    },
  }),
  ...createWeather("weather.entity"),
  ...createClimate("climate.air_conditioner"),
  ...createVacuum("vacuum.robot_vacuum"),
} as const;
