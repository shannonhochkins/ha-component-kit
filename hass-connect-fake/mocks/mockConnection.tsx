import type { HassEntities } from "home-assistant-js-websocket";
import 'jest-localstorage-mock';
import { entities } from './mockEntities';
const timeRegex = /\d{2}:\d{2}/g;
const dateRegex = /20(2[3-9]|[3-9][0-9])-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z/g;
const ENTITIES = JSON.parse(JSON.stringify(entities)  
  // purposely using 2022 here as the regex should work for 9 years ;)
  .replace(dateRegex, '2022-07-23T03:14:01.164Z')
  .replace(timeRegex, '01:01'));

localStorage.setItem('hassTokens', JSON.stringify({
  access_token: "FAKE",
  token_type:"Bearer",
  expires_in: Number.MAX_SAFE_INTEGER,
  hassUrl: "http://fake.com",
  clientId: "http://localhost:6006/",
  expires: Number.MAX_SAFE_INTEGER,
  refresh_token:"FAKE"
}));
// eslint-disable-next-line react-refresh/only-export-components
export const connection = {
  id: 'fake'
}
// eslint-disable-next-line react-refresh/only-export-components
export const mocked = {
  callService: jest.fn(),
  getAllEntities: jest.fn().mockImplementation(() => ENTITIES),
  async getAuth() {
    return {
      expired: false,
    }
  },
  async getConfig() {
    return {
      time_zone: 'Australia/Sydney'
    }
  },
  async createConnection() {
    return connection;
  },
  subscribeEntities: jest.fn().mockImplementation((_: null, subScribeFn: (entities: HassEntities) => void) => {
    subScribeFn(ENTITIES);
  })
}

jest.mock('home-assistant-js-websocket', () => mocked);

import { HassConnect } from '@hakit/core';

// eslint-disable-next-line react-refresh/only-export-components
export const onReady = jest.fn();
// Custom wrapper component for testing
export const TestWrapper = ({ children }: { children: React.ReactNode }) => {
  return <HassConnect onReady={onReady} hassUrl="http://fake.com">
    {children}
  </HassConnect>
}