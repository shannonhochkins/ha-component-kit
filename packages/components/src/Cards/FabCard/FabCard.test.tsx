// important that these are imported first or the mock won't work.
import { TestWrapper, onReady, mocked, connection } from "@hass-connect-fake/mocks/mockConnection";
import { FabCard } from "./";
import { computeDomain } from "@utils/computeDomain";
import { render, waitFor, fireEvent } from "@testing-library/react";
import { snakeCase } from "lodash";
describe("<FabCard />", () => {
  beforeEach(() => {
    onReady.mockClear();
  });
  // Test to check if the component renders correctly
  it("renders FabCard", async () => {
    const { getByTestId } = render(<FabCard entity="light.fake_light_1" service="turnOn" data-testid="fab-card" />, {
      wrapper: TestWrapper,
    });
    await waitFor(() => expect(onReady).toHaveBeenCalledTimes(1));
    const buttonElement = getByTestId("fab-card");
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toMatchSnapshot();
  });

  // Tests for all valid combinations of entity and service props
  const entities = ["light.fake_light_1", "switch.fake_switch"] as const;
  const services = ["turnOn", "toggle", "turnOff"] as const;

  entities.forEach((entity) => {
    services.forEach((service) => {
      it(`renders correctly with entity ${entity} and service ${service}`, async () => {
        const { getByTestId } = render(<FabCard entity={entity} service={service} data-testid="fab-card" />, {
          wrapper: TestWrapper,
        });
        await waitFor(() => expect(onReady).toHaveBeenCalledTimes(1));
        const buttonElement = getByTestId("fab-card");
        expect(buttonElement).toBeInTheDocument();
      });

      it(`triggers service ${service} when clicked with entity ${entity}`, async () => {
        const mockFunction = jest.fn();
        const { getByTestId } = render(<FabCard entity={entity} service={service} onClick={mockFunction} data-testid="fab-card" />, {
          wrapper: TestWrapper,
        });
        await waitFor(() => expect(onReady).toHaveBeenCalledTimes(1));
        const buttonElement = getByTestId("fab-card");
        fireEvent.click(buttonElement);
        expect(mocked.callService).toHaveBeenLastCalledWith(connection, computeDomain(entity), snakeCase(service), undefined, {
          entity_id: entity,
        });
        expect(mockFunction).toHaveBeenCalled();
      });
    });
  });

  it("renders FabCard with a size default of 48", async () => {
    const { getByTestId } = render(<FabCard entity="light.fake_light_1" service="turnOn" data-testid="fab-card" />, {
      wrapper: TestWrapper,
    });
    await waitFor(() => expect(onReady).toHaveBeenCalledTimes(1));
    const buttonElement = getByTestId("fab-card");
    expect(buttonElement.getAttribute("size")).toEqual("48");
  });

  it("should render without an entity", async () => {
    const { getByTestId } = render(<FabCard icon="mdi:cross" data-testid="fab-card" />, {
      wrapper: TestWrapper,
    });
    await waitFor(() => expect(onReady).toHaveBeenCalledTimes(1));
    const buttonElement = getByTestId("fab-card");
    expect(buttonElement).toBeInTheDocument();
  });
});
