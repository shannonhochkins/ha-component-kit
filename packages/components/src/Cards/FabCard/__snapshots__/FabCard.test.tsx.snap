// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FabCard /> renders FabCard 1`] = `
.emotion-0 {
  all: unset;
  position: relative;
  overflow: hidden;
  cursor: pointer;
  text-align: center;
  border-radius: 50%;
  background-color: var(--ha-secondary-background);
  color: var(--ha-300);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);
  -webkit-transition: var(--ha-transition-duration) var(--ha-easing);
  transition: var(--ha-transition-duration) var(--ha-easing);
  transition-property: background-color,box-shadow;
  font-size: 0.7rem;
  height: 48px;
  width: 48px;
}

.emotion-0:hover {
  background-color: var(--ha-secondary-background-hover);
}

<button
  class="emotion-0"
  data-testid="fab-card"
  size="48"
  style="opacity: 1;"
>
  <span />
</button>
`;
